### PR TITLE
Stabilize deque_make_contiguous

### LINKED ISSUE #70929: Tracking Issue for `fn VecDeque::make_contiguous`
This is a tracking issue for `fn make_contiguous` of `VecDeque`.
The feature gate for the issue is `#![feature(deque_make_contiguous)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] include a note in the `VecDeque` struct docs recommending `make_contiguous` for sorting.
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr

### Implementation history
- initial implementation in #69425

### PR DESCRIPTION
**Pull request by jonhoo**
Closes #70929.

/cc @Amanieu

### CODE REVIEW

**Comment by jonhoo** [comment]
(I cc'ed Amanieu after a [brief conversation](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/vecdeque_make_contiguous/near/204421749) with him on Zulip where he told me to post this)

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by jonhoo** [comment]
Gentle ping on this. Should hopefully not be too controversial, and it'd be great to try to get it onto the release train :)

File: library/alloc/src/collections/vec_deque.rs [hunk 1]
```diff
@@ -47,10 +47,17 @@ const MAXIMUM_ZST_CAPACITY: usize = 1 << (64 - 1); // Largest possible power of
 /// push onto the back in this manner, and iterating over `VecDeque` goes front
 /// to back.
 ///
+/// Since `VecDeque` is a ring buffer, its elements are not necessarily contiguous
+/// in memory. If you want to access the elements as a single slice, such as for
+/// efficient sorting, you can use [`make_contiguous`]. It rotates the `VecDeque`
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/vec_deque.rs:52]: Hm -- I would expect that we could implement sorting almost as efficiently on two slices as one? We still have Index/IndexMut impls, so anything a sort would need to do seems like it should be essentially just as fast.

Comment (jonhoo) [reply library/alloc/src/collections/vec_deque.rs:52]: A sort impl for `VecDeque` was submitted in https://github.com/rust-lang/rust/pull/69400#issuecomment-597644295, but was then abandoned in favor of `make_contiguous`. I agree with you that it's probably possible to implement an efficient sort directly on a `VecDeque`, though it's certainly a bit more complicated (and would also add many extra methods). For example, a straightforward heapsort wouldn't work unless you do some tricky wrapping index calculation for child/parent indexing. Or you could go for two heapsorts + a merge (which I think the original PR did), but that is also a bit slower.

Ultimately, I think `make_contiguous` is useful regardless of whether we get sorting. And this doc text can always be updated to reflect added sorting methods if those get added later.

Comment (Mark-Simulacrum) [reply library/alloc/src/collections/vec_deque.rs:52]: :+1: seems reasonable.

**Comment by Amanieu** [comment]
ping for checkboxes @SimonSapin @withoutboats @sfackler

File: library/alloc/src/collections/vec_deque.rs [hunk 2]
```diff
@@ -2228,7 +2231,7 @@ impl<T> VecDeque<T> {
     ///     assert_eq!(slice, &[3, 2, 1] as &[_]);
     /// }
     /// ```
-    #[unstable(feature = "deque_make_contiguous", issue = "70929")]
+    #[stable(feature = "deque_make_contiguous", since = "1.47.0")]
```
Comment (Amanieu) [inline library/alloc/src/collections/vec_deque.rs:2234]: Can you bump this to 1.48.0?

**Comment by jonhoo** [comment]
@Amanieu done! Also merged master for good measure.

**Review by dtolnay — APPROVED** [review]

**Pull request merged**